### PR TITLE
Add m2e support

### LINKED ISSUE #1413: Support for execution during incremental builds with m2eclipse
[M2E](https://www.eclipse.org/m2e/index.html), the Maven integration feature in Eclipse, can optionally support the execution of Maven Plugins during incremental compilation. For that it is necessary to add some metadata to the plugin and also notify via the plexus-build-api about updates of files: https://www.eclipse.org/m2e/documentation/m2e-making-maven-plugins-compat.html

### PR DESCRIPTION
**Pull request by kwin**
This closes #1413

Please **DO NOT FORCE PUSH**. Don't worry about messy history, it's easier to do code review if we can tell what happened after the review, and force pushing breaks that.

Please make sure that your [PR allows edits from maintainers](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/).  Sometimes its faster for us to just fix something than it is to describe how to fix it.

Allow edits from maintainers

After creating the PR, please add a commit that adds a bullet-point under the `[Unreleased]` section of [CHANGES.md](https://github.com/diffplug/spotless/blob/main/CHANGES.md), [plugin-gradle/CHANGES.md](https://github.com/diffplug/spotless/blob/main/plugin-gradle/CHANGES.md), and [plugin-maven/CHANGES.md](https://github.com/diffplug/spotless/blob/main/plugin-maven/CHANGES.md) which includes:

- [ ] a summary of the change
- either
    - [ ] a link to the issue you are resolving (for small changes)
    - [ ] a link to the PR you just created (for big changes likely to have discussion)

If your change only affects a build plugin, and not the lib, then you only need to update the `plugin-foo/CHANGES.md` for that plugin.

If your change affects lib in an end-user-visible way (fixing a bug, updating a version) then you need to update `CHANGES.md` for both the lib and all build plugins.  Users of a build plugin shouldn't have to refer to lib to see changes that affect them.

This makes it easier for the maintainers to quickly release your changes :)

### CODE REVIEW

File: plugin-maven/src/main/java/com/diffplug/spotless/maven/incremental/UpToDateChecker.java [hunk 1]
```diff
@@ -37,4 +38,25 @@ static UpToDateChecker noop(MavenProject project, Path indexFile, Log log) {
...
+		return new UpToDateChecker() {
+
+			@Override
+			public void setUpToDate(Path file) {
+				delegate.setUpToDate(file);
+			}
+
+			@Override
+			public boolean isUpToDate(Path file) {
+				return !buildContext.hasDelta(file.toFile()) || delegate.isUpToDate(file);
```
Comment (nedtwigg) [inline plugin-maven/src/main/java/com/diffplug/spotless/maven/incremental/UpToDateChecker.java:52]: Should this be `&&` instead of `||`?

**Review by nedtwigg — COMMENTED** [review]
Thanks, great PR! I've got one minor nit below, address that and then I'd be happy to merge. @lutovich do you want to take a look? No worries if you're too busy, if I don't hear from you in a week I'll assume you're okay with it.

File: plugin-maven/src/main/java/com/diffplug/spotless/maven/AbstractSpotlessMojo.java [hunk 2]
```diff
@@ -88,6 +89,9 @@ public abstract class AbstractSpotlessMojo extends AbstractMojo {
 	@Component
 	private ResourceManager resourceManager;
 
+	@Component
+	protected BuildContext buildContext;
```
Comment (lutovich) [inline plugin-maven/src/main/java/com/diffplug/spotless/maven/AbstractSpotlessMojo.java:93]:
```suggestion
	private BuildContext buildContext;
```

**Review by lutovich — APPROVED** [review]
LGTM! I added one nit suggestion and have two questions just to better understand the change (I have no experience with m2e):
 1. Is my understanding correct that a simple non-Eclipse maven build will inject a [DefaultBuildContext](https://github.com/sonatype/sisu-build-api/blob/444c1afff47b8f5d355f9190b3fb1501f907ab77/src/main/java/org/sonatype/plexus/build/incremental/DefaultBuildContext.java#L39) that will not affect up-to-date checking?
 2. What `BuildContext` will be injected when a build happens in Eclipse?

**Comment by lutovich** [comment]
Looks like there's a compilation error:
```
> Task :plugin-maven:compileJava FAILED
/home/circleci/project/plugin-maven/src/main/java/com/diffplug/spotless/maven/incremental/UpToDateChecker.java:52: error: no suitable method found for hasDelta(Path)
				if (buildContext.hasDelta(file)) {
				                ^
    method BuildContext.hasDelta(String) is not applicable
      (argument mismatch; Path cannot be converted to String)
    method BuildContext.hasDelta(File) is not applicable
      (argument mismatch; Path cannot be converted to File)
    method BuildContext.hasDelta(List) is not applicable
      (argument mismatch; Path cannot be converted to List)
1 error
```

**Comment by kwin** [comment]
> LGTM! I added one nit suggestion and have two questions just to better understand the change (I have no experience with m2e):
> 
> 1. Is my understanding correct that a simple non-Eclipse maven build will inject a [DefaultBuildContext](https://github.com/sonatype/sisu-build-api/blob/444c1afff47b8f5d355f9190b3fb1501f907ab77/src/main/java/org/sonatype/plexus/build/incremental/DefaultBuildContext.java#L39) that will not affect up-to-date checking?

Yes, compare also with https://github.com/codehaus-plexus/plexus-build-api/blob/master/README.md#default-implementation

> 2. What `BuildContext` will be injected when a build happens in Eclipse?

The one from https://github.com/eclipse-m2e/m2e-core/blob/d4c976bb9c8d0aabb029030c3c4f3430af8a8c9e/org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/builder/plexusbuildapi/EclipseBuildContext.java

File: plugin-maven/src/main/java/com/diffplug/spotless/maven/AbstractSpotlessMojo.java (hunk 2 above)
Comment (kwin) [reply plugin-maven/src/main/java/com/diffplug/spotless/maven/AbstractSpotlessMojo.java:93]: Not possible as accessed by `SpotlessApplyMojo`.

**Review by lutovich — APPROVED** [review]

**Comment by nedtwigg** [comment]
Released in `plugin-maven 2.29.0`.

**Pull request merged**